src/lib/doctrine-dbal-bulk/src/Flow/Doctrine/Bulk/Dialect/MySQLDialect.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\Doctrine\Bulk\Dialect;
+
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use Flow\Doctrine\Bulk\BulkData;
+use Flow\Doctrine\Bulk\Columns;
+use Flow\Doctrine\Bulk\Exception\RuntimeException;
+use Flow\Doctrine\Bulk\TableDefinition;
+
+final class MySQLDialect implements Dialect
+{
+    private AbstractPlatform $platform;
+
+    public function __construct(AbstractPlatform $platform)
+    {
+        $this->platform = $platform;
+    }
+
+    /**
+     * @param TableDefinition $table
+     * @param BulkData $bulkData
+     * @param array{
+     *  skip_conflicts?: boolean,
+     *  constraint?: string,
+     *  conflict_columns?: array<string>,
+     *  update_columns?: array<string>
+     * } $insertOptions $insertOptions
+     *
+     * @return string
+     */
+    public function prepareInsert(TableDefinition $table, BulkData $bulkData, array $insertOptions = []) : string
+    {
+        if (\array_key_exists('conflict_columns', $insertOptions)) {
+            return \sprintf(
+                'INSERT INTO %s (%s) VALUES %s ON DUPLICATE KEY UPDATE %s',
+                $table->name(),
+                $bulkData->columns()->concat(','),
+                $bulkData->toSqlPlaceholders(),
+                (\array_key_exists('update_columns', $insertOptions) && \count($insertOptions['update_columns']))
+                    ? $this->updatedSelectedColumns($insertOptions['update_columns'], $bulkData->columns())
+                    : $this->updateAllColumns($bulkData->columns())
+            );
+        }
+
+        if (\array_key_exists('skip_conflicts', $insertOptions) && $insertOptions['skip_conflicts'] === true) {
+            return \sprintf(
+                'INSERT INTO %s (%s) VALUES %s ON DUPLICATE KEY UPDATE %4$s=%4$s',
+                $table->name(),
+                $bulkData->columns()->concat(','),
+                $bulkData->toSqlPlaceholders(),
+                current($bulkData->columns()->all())
+            );
+        }
+
+        return \sprintf(
+            'INSERT INTO %s (%s) VALUES %s',
+            $table->name(),
+            $bulkData->columns()->concat(','),
+            $bulkData->toSqlPlaceholders()
+        );
+    }
+
+    /**
+     * @param TableDefinition $table
+     * @param BulkData $bulkData
+     * @param array{
+     *  primary_key_columns?: array<string>,
```

Comment:
not used?

Reply:
Unfortunately, mysql doesn't support such feature afaik when updating on conflict, I've fixed annotations and remove not used functions.